mem-cache: Add unit test for LFU RP

odanrc opened the pull request:
As side effects:
- We now verify that the repl data is not null

odanrc committed:
mem-cache: Add unit test for LFU RP

As side effects:
- We now verify that the repl data is not null

Signed-off-by: Daniel R. Carvalho <odanrc@yahoo.com.br>

erin-le commented on src/mem/cache/replacement_policies/lfu_rp.test.cc:
```diff
@@ -0,0 +1,286 @@
+
+/// Test that when all entries have been reset the same entry will always
+/// be the one victimized
+TEST_F(LFURPVictimizationTestF, GetVictimAllReset)
```
Is there supposed to be more to this test? The comment mentions checking that the same entry is "always" victimized, so I was wondering if there were other conditions you also wanted to check.

odanrc replied:
Not really; I reused the tests/documentation from the other test, making the necessary changes. In theory it would make sense to run other scenarios (e.g., reset entry[0] last to make sure it is still picked), but at this point it feels like a guessing game of which bugs could occur.

erin-le reviewed (commented):
@odanrc: Hello, thank you for this PR. The tests generally look good to me, but it looks like this PR is failing the CI tests because of the dynamic_pointer_cast you have on lines 51 and 52. Could you take a look? Thank you!

odanrc commented on src/mem/cache/replacement_policies/lfu_rp.test.cc:
```diff
@@ -0,0 +1,286 @@
+        const std::shared_ptr<gem5::replacement_policy::ReplacementData>
+            repl_data) const
+    {
+        return nullptr !=
```
The compiler complains: ´source type is not polymorphic´

The issue is that the empty struct/base type ReplacementData is not polymorphic, and I would have to at least declare its destructor virtual. I don´t think it is sane to do this change in the source because of a test, so I have removed this validation from the test.

odanrc committed:
mem-cache: Remove LFU repl data validation

Since the base class is not polymorphic, and we don't want it
to be, we cannot rely on dynamic_cast to validate the type of
the entry created. As such, we are forced to remove this
validation.

Signed-off-by: Daniel R. Carvalho <odanrc@yahoo.com.br>

odanrc commented:
@erin-le , are you able to see what is the clang-format error? I run it locally, and it complains about nothing. I am using clang-format 15.0.7. What is the version being used for the verification?

odanrc commented:
> @erin-le , are you able to see what is the clang-format error? I run it locally, and it complains about nothing. I am using clang-format 15.0.7. What is the version being used for the verification?

I rebased on top of develop and uploaded again, and the error is gone. I didn't make any changes to the code...

erin-le reviewed (approved):
LGTM. I'm not 100% sure about this, but I think the clang-format test sometimes fails if the `.clang-format` file is updated on `develop` after the PR branch has been created, and the `develop` branch and PR end up checking for different formatting. The last update to the `.clang-format` file was 4 months ago, however, so I don't know if that's the case here.
